keps/sig-cluster-lifecycle/wgs/2141-instance-specific-componentconfig/README.md

```diff
@@ -0,0 +1,664 @@
+from the configuration workflow. This is a great idea in theory, because it
+solves the problem by removing work for users, but it may be difficult in
+practice. For example, on a machine with multiple network cards, which IP should
+the kubelet use?
+
+This solution is orthogonal and does not prevent us from finding ways to
+auto-detect instance specific config in the future. It is worth noting that
+the _possibility_ of someday being able to auto-detect these has led to some
+hesitation to make a firm decision today. We have delayed for too long already,
+ and a firm decision is needed to move forward.
+
+### Risks and Mitigations
+
+For users deploying Pod templates, this solution can make the template more
+verbose. In the future, we can explore more elegant ways of generating files
+with instance-specific parameters correctly substituted. This proposal argues
+that the extra verbosity is worth the benefit of a consistent, versioned API for
+configuring core components.
+
+## Design Details
+
+The canonical flag for providing a ComponentConfig file to a component is
+`--config`. This KEP proposes that the canonical flag for providing an
+instance-specific config file be `--instance-config`.
+
+As with sharable ComponentConfig parameters, command line flags for
+instance-specific parameters should continue to function and take precedence
+over the final merged values from the config file so that backward compatibility
+is maintained.
+
+### Configuration merge example
+
+The merge between both YAML configuration files is provided by the
+[Strategic Merge Patch](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/strategic-merge-patch.md)
+library.
+
+The example below is a sharable Kubelet configuration YAML file.
+
+```yaml
+apiVersion: kubelet.config.k8s.io/v1beta1
+authorization:
+  mode: Webhook
+  webhook:
+    cacheAuthorizedTTL: 0s
+    cacheUnauthorizedTTL: 0s
+clusterDNS:
+- 10.96.0.10
+clusterDomain: cluster.local
+cpuManagerReconcilePeriod: 0s
+evictionHard:
+  imagefs.available: 0%
+  nodefs.available: 0%
+  nodefs.inodesFree: 0%
+evictionPressureTransitionPeriod: 0s
+featureGates:
+  DynamicKubeletConfig: true
+healthzBindAddress: 127.0.0.1
+healthzPort: 10248
+kind: KubeletConfiguration
+rotateCertificates: true
+staticPodPath: /etc/kubernetes/manifests
+```
+
+This is the instance-configuration YAML, which is going to be merged into the sharable:
+
+```yaml
+address: 127.0.0.1
+apiVersion: kubelet.config.k8s.io/v1beta1
+clusterDNS:
+- 10.96.0.11
+evictionHard:
+  imagefs.available: 2%
+featureGates:
+  TaintBasedEvictions: true
+kind: KubeletConfiguration
+```
+
+The final patched YAML, used to unmarshal into the KubeletConfiguration object is:
+
+```yaml
+address: 127.0.0.1
+apiVersion: kubelet.config.k8s.io/v1beta1
+authorization:
+  mode: Webhook
+  webhook:
+    cacheAuthorizedTTL: 0s
+    cacheUnauthorizedTTL: 0s
+clusterDNS:
+- 10.96.0.11
```

Comment:
how do you know this should replace and not merge? I mean technically it is clear. But how do you know the user does not want to merge?

Reply:
The user will not have control over each field, at least on this point, the choice is still bound to the implementation tagging.